src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -552,41 +556,141 @@ def calculate_dense(self):
     def expectation(self, state, normalize=False):
         return Hamiltonian.expectation(self, state, normalize)
 
-    def expectation_from_samples(self, freq, qubit_map=None):
-        terms = self.terms
-        for term in terms:
+    def _exp_from_circuit(
+        self, circuit: "Circuit", qubit_map: dict, nshots: int = None
+    ):
+        """
+        Calculate the expectation value from a circuit.
+        This allows for non diagonal observables. Each term of the observable is
+        treated separately, by measuring in the correct basis and re-executing the
+        circuit.
+
+        Args:
+            circuit (Circuit): input frequencies.
+            qubit_map (dict): qubit map.
+
+        Returns:
+            (float): the calculated expectation value.
+        """
+        from qibo import Circuit, gates
+
+        if nshots is None:  # pragma: no cover
+            nshots = 1000
```

Comment:
Why not use `nshots=1000` directly as the default? I think we are doing the same in `Circuit.execute`.

I understand that this value would be ignored if frequencies are passed in the `expectation_from_samples` function, which probably motivates the use of `None`, but ignoring `1000` should also be fine (up to debate).

Reply:
Specifically here, it is None because it depends on `expectation_from_samples` which defaults to None as well. But yes, we could just directly set it to 1000, no problem, I was not just sure whether this would cause any confusion, even though in the docstring is explicitely mentioned.

In any case, probably the best solution would be to completely separate the two functions and expose them directly to the user, as you suggested below.